### DISCUSSION TITLE
v2.0.0 : Multicontainer Pre-Release

### DISCUSSION THREAD

**Discussion by PoeppingT** [awslabs/aws-saas-boost#206] [Announcements]
Hello curious reader! These past few months we've been hard at work on the v2.0.0 release of SaaS Boost, and our first stable release candidate is now ready for preview! This release candidate includes a multitude of changes to the internals of SaaS Boost to enable the highly requested Multicontainer support feature!

Download the source code here: https://github.com/awslabs/aws-saas-boost/releases/tag/v2.0.0-rc1

Or clone the source code into your local git repository by running 
```
git fetch
git checkout tags/v2.0.0-rc1
```

To test it out, create a new SaaS Boost environment using this release candidate. When logging into the administrative UI you will notice a new Application configuration page which allows you to define multiple _services_ to be installed into separate compute environments representing the internal structure of your application. 

This is your chance to help us kick the tires and provide feedback. Please respond on the GitHub discussion for the v2.0.0 release with comments, questions, and feedback. Updated documentation and bug fixes are coming, and new release candidates will be cut periodically as we continue to build changes towards a final v2.0.0 release later this quarter. Stay tuned!

**Comment by tieujason330** [comment]
Hey, I'm currently trying to use RC0. When creating a new tenant, it's stuck in a `Validating` state due a missing tier config.
I see that the warning `Missing tier configuration for service 'xxx' tier 'default '`.

In `OnboardingFormComponent.js`, there's an extra empty space after `'default'` in the `initialValues` obj
```
  const initialValues = {
    name: '',
    tier: tier || 'default ',
    subdomain: '',
    billingPlan: '',
    hasBilling: hasBilling,
    hasDomain: hasDomain,
  }
```

I also notice that there's different tiers available when onboarding a new tenant (silver, gold ,etc) but I don't see how to create/define those tiers.

**Reply by PoeppingT** [reply]
@tieujason330 Thanks for the feedback! You're right, it looks like that initial value for tier has an extra space, causing validation failure in the onboarding validation phase. I've created a bug report, we'll build a fix for this into RC1: https://github.com/awslabs/aws-saas-boost/issues/208

There isn't any way in the UI yet to create or modify existing tiers, but you can do so through the API. On the dev team the tool we use is Postman, but you can also simply access the API by going to API Gateway in your console and sending requests that way.

Thanks for checking out RC0! We look forward to more feedback as you keep testing things out.

**Reply by tieujason330** [reply]
Thanks for the tip on dealing w/ adding tiers.

I ran into some other problems

Two ALB listeners are trying to be created on port 80, so the CFT stack fails due to this conflict
Fixed it w/ these changes (similar to what was done pre-2.0.0)
```
diff --git a/functions/onboarding-stack-listener/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/OnboardingStackListener.java b/functions/onboarding-stack-listener/src/main/java/com/amazon/aws/partners/saasfactory/saa
sboost/OnboardingStackListener.java
index 044149d..53bbaee 100644
--- a/functions/onboarding-stack-listener/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/OnboardingStackListener.java
+++ b/functions/onboarding-stack-listener/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/OnboardingStackListener.java
@@ -199,7 +199,7 @@ public class OnboardingStackListener implements RequestHandler<SNSEvent, Object>
                                                 "consoleUrl", AwsResource.LOAD_BALANCER.formatUrl(AWS_REGION, "sb-" + SAAS_BOOST_ENV + "-tenant-" + tenantId.split("-")[0]))
                                         );
                                     } else if ("AWS::ElasticLoadBalancingV2::Listener".equals(resourceType)) {
-                                        if ("HttpListener".equals(logicalId)) {
+                                        if ("HttpListener".equals(logicalId) || "RedirectToHttpsListener".equals(logicalId)) {

diff --git a/resources/tenant-onboarding.yaml b/resources/tenant-onboarding.yaml
index 13b1b08..f24a290 100644
--- a/resources/tenant-onboarding.yaml
+++ b/resources/tenant-onboarding.yaml
@@ -46,7 +46,8 @@ Conditions:
   HasHostedZone: !Not [!Equals [!Ref HostedZoneId, '']]
   HasSubDomainName: !Not [!Equals [!Ref TenantSubDomain, '']]
   HasCertificate: !Not [!Equals [!Ref SSLCertificateArn, '']]
-  CreateSubDomainAlias: !And
+  NoCertificate: !Not [Condition: HasCertificate]
+  CreateSubDomainAlias: !And
     - !Condition HasDomainName
     - !Condition HasHostedZone
     - !Condition HasSubDomainName
@@ -299,6 +300,7 @@ Resources:
       Port: 80
       Protocol: HTTP
   HttpListener:
+    Condition: NoCertificate
     Type: AWS::ElasticLoadBalancingV2::Listener
```

Then, I'm trying to add a RDS extension to one of my services, and get the following exception. Looks like the onboarding svc is attempting to parse the `bootstrapFilename` field even though it's optional and may be null.
```
CloudFormation template parameter RDSBootstrap is NULL: java.lang.RuntimeException
java.lang.RuntimeException: CloudFormation template parameter RDSBootstrap is NULL
	at com.amazon.aws.partners.saasfactory.saasboost.OnboardingService.handleOnboardingBaseProvisioned(OnboardingService.java:953)
	at com.amazon.aws.partners.saasfactory.saasboost.OnboardingService.handleOnboardingEvent(OnboardingService.java:352)
	at jdk.internal.reflect.GeneratedMethodAccessor3.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
```

tried fixing it w/ the below although i'm running into more issues. 
```
diff --git a/services/onboarding-service/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/OnboardingService.java b/services/onboarding-service/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/OnboardingServi
ce.java
index 1b88131..d815450 100644
--- a/services/onboarding-service/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/OnboardingService.java
+++ b/services/onboarding-service/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/OnboardingService.java
@@ -472,7 +472,7 @@ public class OnboardingService {
                 String tenantId = (String) insertedTenant.get("id");
                 onboarding.setTenantId(UUID.fromString(tenantId));
                 onboarding = dal.updateOnboarding(onboarding);
                 // Assign a CIDR block to this tenant to use for its VPC
                 try {
                     dal.assignCidrBlock(tenantId);
@@ -889,7 +889,9 @@ public class OnboardingService {
                             dbDatabase = (String) database.get("database");
                             dbUsername = (String) database.get("username");
                             dbPort = (Integer) database.get("port");
-                            dbBootstrap = (String) database.get("bootstrapFilename");
+                            if (database.get("bootstrapFilename") != null) {
+                                dbBootstrap = (String) database.get("bootstrapFilename");
+                            }
                             dbPasswordRef = (String) database.get("passwordParam");
                         }
```

Now I'm running into CFT stack errors since it looks like multiple stacks for the same service are attempting to be created. Do you have any tips on how I can debug/iterate quicker? Right now, I'm relying on digging through the CloudWatch logs and updating the Saas Boost deployment w/ new changes so it takes a while for everything to get updated.

**Reply by PoeppingT** [reply]
It looks like you've got the troubleshooting part down: digging through CloudWatch logs is the first place to look to find bugs. If you're just replacing a single service, you can use the `update-service.sh` scripts in each service directory to simply replace that single lambda's worth of code. There are similar scripts for the lambda layers and CloudFormation resources (in the `resources/` directory).

**Reply by tieujason330** [reply]
I've been making progress on getting my app to work. one issue i'm running into is that the S3 pipeline bucket doesn't have any of the imagedef.jsons. From digging around, the `workload-deploy` lambda looks for all tenants that match the filter `tenants?status=provisioned`. is it possible that at the time when the codepipeline tenant resources get added, the status isn't `provisioned` yet? this is causing my CodePipeline workflows to fail because the Source action never starts (nothing in the s3 buckets). Or can you help me get some insight on what else could be causing my issue? I haven't made any local changes dealing with codepipelines.

some relevant logs:
tenant resources updated w/ SERVICE_KIBANA_CODE_PIPELINE` via onboard-stack-app-listener` lambda
```
2022-04-20 23:42:18.592 | 2022-04-20 23:42:18.592 5df00c02-ba96-4ad7-bccb-8bb04cf02b0f INFO  OnboardingAppStackListener - Publishing update tenant resources event for tenant 563e5800-3086-4af8-99f6-022af4c8a339 SERVICE_KIBANA_CODE_PIPELINE sb-jasondev-tenant-563e5800-kibana
```

`WorkloadDeploy` lambda logs error saying `SERVICE_KIBANA_CODE_PIPELINE` cannot be found
```
2022-04-20 23:54:25.957 | 2022-04-20 23:54:25.957 63df3fcc-8513-472f-b8c7-6fd3c14fbfd9 ERROR WorkloadDeploy - Can't find CodePipeline resource SERVICE_KIBANA_CODE_PIPELINE for tenant 563e5800-3086-4af8-99f6-022af4c8a339
```

-------
Also, here some small changes that i've locally made that may be useful:
- `OnboardStackListener.java::filter()` currently filters for `AWS::CloudFormation::Stack` and stack names that matches the tenant id, but this is problematic if we add a nested`AWS::CloudFormation::Stack` (in my scenario, it's AWS Elasticsearch) since it seems to inherit the parent stack name. i changed it to check against the logical resource id to prevent my nested stack's `CREATE_COMPLETE` event from triggering anything downstream
- all tenant related lambda names are prefixed `sb-<env>-tenants-` except for `sb-<ent>-tenant-events`. i think it'd keeping it the same prefix as the rest would be best since `./update-service.sh` won't update the events lambda (took me a while to figure out why changes to the events lambda weren't working)
- codepipline actions are missing, resulting in a cft warning `Did not have IAM permissions to process tags on AWS::CodePipeline::Pipeline resource.`
   - codepipeline:TagResource
   - codepipeline:UntagResource

**Reply by PoeppingT** [reply]
That `WorkloadDeploy` function will only run after the entire stack has completed: it's only run once the `handleOnboardingProvisioned` event is received, which only happens once all `onboarding-app` stacks complete.

The `SERVICE_KIBANA_CODE_PIPELINE` key that the `WorkloadDeploy` function is looking for in this case should be a part of the Tenant resources, which should be updated in the Tenant service in response to events sent by the `OnboardingAppStackListener` function. I see you were making changes to the Tenant events lambda: this is the lambda that should be updating that Tenant record with the Tenant resources [fired off by the `OnboardingAppStackListener`](https://github.com/awslabs/aws-saas-boost/blob/multi-container/functions/onboarding-app-stack-listener/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/OnboardingAppStackListener.java#L183-L186):
```
if ("AWS::CodePipeline::Pipeline".equals(resource.resourceType())) {
    String codePipeline = resource.physicalResourceId();
    // The resources collection on the tenant object is Map<String, Resource>
    // so we need a unique key per service code pipeline. We'll prefix the
    // key with SERVICE_ and suffix it with _CODE_PIPELINE so we can find
    // all of the tenant's code pipelines later on by looking for that pattern.
    String key = serviceNameResourceKey(serviceName, AwsResource.CODE_PIPELINE.name());
    LOGGER.info("Publishing update tenant resources event for tenant {} {} {}", tenantId,
            key, codePipeline);

    tenantResources.put(key, Map.of(
            "name", codePipeline,
            "arn", AwsResource.CODE_PIPELINE.formatArn(partition, AWS_REGION, accountId,
                codePipeline),
            "consoleUrl", AwsResource.CODE_PIPELINE.formatUrl(AWS_REGION, codePipeline)
    ));
...
if (!tenantResources.isEmpty()) {
    // The update tenant resources API call is additive, so we don't need to pull the
    // current tenant object ourselves.
    Utils.publishEvent(eventBridge, SAAS_BOOST_EVENT_BUS, EVENT_SOURCE,
            "Tenant Resources Changed",
            Map.of("tenantId", tenantId, "resources", Utils.toJson(tenantResources))
    );
}
```

You should see some logs in `sb-<env>-tenant-events` that indicate to you what went wrong, or at least will show you that the timing got messed up between these events.

Thanks for the suggestion on bugfixes, if you have the time it would be useful to us for you to open issues against the repository so we can work them into our normal bug triaging process.

**Reply by tieujason330** [reply]
Ah thanks for the explanation, it helped me figure out what was wrong on my end.

And sounds good, I'll open issues that I come across, thanks.

**Reply by brtrvn** [reply]
@tieujason330 Make sure your ElasticSearch template is a nested stack on the `tenant-onboarding-app.yaml` stack, not on the tenant-onboarding.yaml stack.

Looks like a bug that the Lambda for the tenant event handler function is not named the same as the rest of the tenant microservice functions. I'll get a fix in for that and then the `update_service.sh` script should work as intended.

**Reply by tieujason330** [reply]
@brtrvn is there a particular reason why ElasticSearch should be a nested stack in the `apps` stack? I nested it within `tenant-onboarding.yaml` since I wanted to pass along elasticsearch info to all my containers (they all rely on Elasticsearch in some way). I was able to do this by storing relevant ElasticSearch info in the `attributes` map when updating the tenant on the stack `CREATE_COMPLETE` event. Feel free to let me know if there could be a better way of handling this, thanks!

**Reply by brtrvn** [reply]
Currently, the isolation model of SaaS Boost is silo based isolation -- meaning every tenant would get their own ElasticSearch cluster. I think you are aware of this, just restating it.

When we extended SaaS Boost to support multiple compute environments (aka services), they are treated as separate entities -- just as you'd expect in a micro services design. So, if `Service A` has a database, then `Service B` shouldn't have access to that database other than through the API contract of Service A.

You can certainly craft a solution to get around this -- as you're doing -- but you're running into errors and challenges because SaaS Boost isn't designed to work this way.

As designed, I'd suggest that you have an _internal_ service that fronts ES and your other containers delegate ES work to that internal service's API to support features backed by your ES cluster.

**Reply by tieujason330** [reply]
Thanks for the insight, I'll review how I integrated ES and aim to fit the micro services design to keep things consistent.

**Comment by feiszli** [comment]
Hi, I'm playing around with SaaS Boost as a way to set up our SaaS. The single container limitation was a hard stop for us, so was excited to see this. Looking at the pre-release, it looks like there are a couple of limitations that will still prevent us from using SaaS Boost. I just want to see if I have it right:

1. We currently deploy our app via docker-compose. It looks like we can't use docker-compose to deploy multi-container, right?
2. We expose multiple ports on our server. It looks like we can only expose one port per container, is this correct?
3. We have a couple of non-http services. It looks like everything needs to be over http, right?

**Reply by PoeppingT** [reply]
Hi @feiszli , thanks for looking at SaaS Boost! Glad to hear you're excited to try working with our multi-container offering. I'll take a shot at answering your questions:
1. At the moment docker-compose is not usable alongside SaaS Boost. SaaS Boost and docker-compose solve a similar problem: both provide a way to define your application into multiple services and eventually deploy it to AWS, but only SaaS Boost offers the SaaS-specific services as part of the same offering. We are already aware internally that some customers may want to install their own application their own way, and are working to see how we can redefine SaaS Boost to provide for that. 
2. You're right, currently we only support one exposed port per container. 
3. You are correct, currently services are restricted to http/https. While the ECS exposed ports are listening on TCP in SaaS Boost, with the [only other option being UDP](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-taskdefinition-containerdefinitions-portmappings.html#cfn-ecs-taskdefinition-containerdefinition-portmappings-sourcevolume), the Application Load Balancer only allows for ingress on ports 80 and 443.

Don't let this stop you from giving it a shot, however! Your feedback is always incredibly useful and any more opportunities you have to kick the tires are appreciated.

**Reply by brtrvn** [reply]
Additional thoughts:

The Docker Compose `docker-compose.yml` file is essentially replaced by the application config file in SaaS Boost. You can see what your app config looks like by issuing a `GET` against your public API instance `/settings/config` resource. We give you a user-friendly UI to create the app config, but you can certainly craft your own JSON and `PUT` it against `/settings/config`. Of course, it must be in a valid form that SaaS Boost understands.

Please submit a feature request to the Issues list for multiple port support and explain your use case so we can think through it. Adding multiple port support is pretty straight-forward from a docker standpoint. The hard part is going to be how that maps to the load balancer. As PoeppingT pointed out, the ALB only works with HTTP/S ports. If these other ports are used for internal features to support additional services in your application, you could deploy them as non-public services in SaaS Boost.

You shouldn't be limited to HTTP/S for "private" services. At runtime, your services will have environment variables set for each of the private services you defined in SaaS Boost following a naming convention of `SERVICE_${SERVICE_NAME}_HOST` and `SERVICE_${SERVICE_NAME}_PORT` How you choose to interact with that host over that TCP port is up to your code. So called private services are not fronted by an ALB, but they are load balanced by ECS Service Discovery.

**Comment by PoeppingT** [comment]
Thanks everyone for watching this discussion! The v2.0.0-rc1 release candidate has just been cut, containing fixes for bugs called out by our community in this discussion.

View the release notes here: https://github.com/awslabs/aws-saas-boost/releases/tag/v2.0.0-rc1

**Reply by tieujason330** [reply]
i'm seeing this error when trying out `v2.0.0-rc1` w/o any local changes. i've tried updating my existing env and creating a new env but this error continues to show up.
```
Installation Error: Circular dependency between resources: [InvokeClearLoggingBucket, billing, quota, ClearBucket, privateapi, tenant, settings, publicapi, InvokeClearAccessLogsBucket, core, InvokeClearResourcesBucket, Resources, metrics, SSMParamResourcesBucket, web, metering, onboarding, InvokeClearCodePipelineBucket, InvokeClearWebsiteBucket, InvokeClearAthenaBucket, ClearBucketExecutionRole, user] (Service: CloudFormation, Status Code: 400, Request ID: fb74cb16-d98d-44b7-a490-109c357cb2e2)
```

**Reply by PoeppingT** [reply]
Thanks for calling this out @tieujason330 , we are looking into the issue.

**Reply by ilove2git** [reply]
failed with same issue when installing v2.0.0-rc1

**Reply by PoeppingT** [reply]
Thanks for the updates, all! A fix has been published and the tag has been moved. You should not be able to install a new SaaS Boost environment using `v2.0.0-rc1`. Please let us know if you find any more issues.

**Reply by tieujason330** [reply]
@PoeppingT is this issue (https://github.com/awslabs/aws-saas-boost/issues/74) resolved for v2.0.0? i have an existing hosted zone that exists before creating a saas-boost env, but saas-boost still creates it's own hosted zone

**Reply by PoeppingT** [reply]
Yes @tieujason330 (#74) should be fixed by https://github.com/awslabs/aws-saas-boost/commit/dbb7804. Are you using the latest version of the multi-container branch?

**Reply by tieujason330** [reply]
thanks, i see that it's looking for the comment specifically being `<domain> Public DNS zone`. so does that mean it'll only officially support re-using existing Hosted Zones created by saas-boost? i can manually update my other hosted zone (created by Route53 Registrar) to match that pattern but it'd be nice to be able to specify it somehow. it's not a big deal though

**Reply by PoeppingT** [reply]
That's a good callout. Please create a bug against the repository with a description of the problem and your suggested solution so we can track this bug report.

**Reply by vikasagaws** [reply]
Hi Team, I am getting the below error while trying to install using v2.0.0-rc1 - 

**Installation Error: Error with CloudFormation stack sb-devrc2. Check the events in the AWS CloudFormation Console**

Any pointers? @PoeppingT

**Reply by PoeppingT** [reply]
Hey @vikasagaws , to give you direct advice we would need to know what went wrong, which you can find by looking into the failure events for your base stack (titled `sb-devrc2` in this case) in your CloudFormation console. By scrolling down you should see each resource owned by the CloudFormation stack and all their state changes. There will be one with red text describing failure, and it may be a child stack!

**Comment by kaypee90** [comment]
Is there any update on Multicontainer support? Has it been released? Are there any docs for the feature?

**Reply by brtrvn** [reply]
@kaypee90 multi container support has been available for over a year. Please use the latest release and you can follow the getting started guide for instructions.
